include/oneapi/dpl/experimental/kt/internal/esimd_radix_sort_kernels.h

```diff
@@ -644,10 +639,7 @@ struct __radix_sort_onesweep_kernel
                         _GlobOffsetT, __bin_width, __dpl_esimd::__ens::lsc_data_size::default_size,
                         __dpl_esimd::__ens::cache_hint::uncached, __dpl_esimd::__ens::cache_hint::cached>(
                         __p_prev_group_hist + __local_tid * __bin_width);
-                    // Software barrier is used in order to motivate the compiler
-                    // to generate memory operations in an order which results in better performance
-                    // TODO: check if it is still necessary.
-                    __dpl_esimd::__ns::fence<__dpl_esimd::__ns::fence_mask::sw_barrier>();
+                    __dpl_esimd::__ns::fence<__dpl_esimd::__ns::memory_kind::global>();
```

Comment:
Let me think about it since I do not have clear understanding of why there is a hang after removal of `sw_barrier` and why the introduced barrier helps... One confusing fact is that the fence is applied on group-scope by default (i.e. it guarantees global memory ops visibility within a group), however it is a cross-group synchronization section.

I was able to reproduce the hang while removing the fence. However, it does not reproduce when I replace it with:
`__dpl_esimd::__ns::fence<__dpl_esimd::__ns::memory_kind::local>();`

Note that this fence above:
https://github.com/oneapi-src/oneDPL/pull/1790/files#diff-b0f1146f4a5f7425f04ba4973e688c2c834dc1c4f7563ecb95a3e937bab81c1bR601-R603
guarantees that the global histogram and histograms local for each group are observable by other groups. This should be enough to avoid hanging. In the worst case, a group "N" can scan all the previous histograms.

Reply:
I've gathered the kernels compiled in asm format for Intel® Data Center GPU Max:
 - [sw_fence.txt](https://github.com/user-attachments/files/16748126/sw_fence.txt)
 - [local_fence.txt](https://github.com/user-attachments/files/16748127/local_fence.txt)
 - [no_fence.txt](https://github.com/user-attachments/files/16748128/no_fence.txt)

The major difference between `sw_fence` and `no_fence` is the following:

```asm
_L_k43_10_preHeader:
BB_9:
// B018: Preds:{B017, B018},  Succs:{B018}
(W)     jmpi                                 _L_k43_10_preHeader                                     // $487
```
It looks like an plain endless loop.

I suggest changing the barrier to the local one, guarding it with a macro to turn it off after the next minor version of icpx, and adding a comment. I will create a bug ticket in the meantime.

Reply:
Thanks for taking a deeper look and clarifying on the barriers. I agree that it should not be needed.

I have switched to the local barrier, added a compiler check macro, and added a comment.